connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java

```diff
@@ -489,7 +489,17 @@ boolean isCycle(String topic) {
         String source = replicationPolicy.topicSource(topic);
         if (source == null) {
             return false;
-        } else if (source.equals(sourceAndTarget.target())) {
+        }
+
+        // Fix for https://issues.apache.org/jira/browse/KAFKA-9914
+        final boolean condition;
+        if (replicationPolicy instanceof IdentityReplicationPolicy) {
```

Comment:
This seems a bit brittle. Users could implement their own `ReplicationPolicy` that behaves like `IdentityReplicationPolicy` and this would not catch it. Can we detect what to do by making calls on the `replicationPolicy` instead?

Reply:
So clearly the `IdentityReplicationPolicy` cannot detect cycles, and this is documented https://github.com/apache/kafka/blob/2edaf9ddd30a6417c34f5bce6894a5340632ff1e/docs/upgrade.html#L112

So, would it make sense to move this `isCycle()` method to `ReplicationPolicy` interface (changing that interface, which is marked `evolving`), and have separate implementation for `isCycle` per replication policy, or which kind of approach do you see fitting best here?

Reply:
Currently `MirrorSourceConnector.isCycle` is only called from `MirrorSourceConnector.shouldReplicateTopic`, which is the method that needs to make decision based on `ReplicationPolicy` used. However I'm out of ideas how to detect without adding some methods into `ReplicationPolicy` interface, and modifying the interface feels like somewhat bigger effort.